[lint] Spot errors in the lint flow that we weren't expecting

rswarbrick opened the pull request:
If something goes wrong and the lint tool returns a nonzero exit
status (causing edalize/fusesoc to complain), we normally want to
squash that error message because it's something we have other
warnings about anyway.

However, we *do* want to see the message if something more dramatic
has happened. For example, we weren't seeing syntax error messages
from a very confused Verible, meaning that we reported no lint errors
when in fact the tool had died completely.

This PR will obviously fail CI at the moment (that's sort of the point!). We'll either have to waive otp_ctrl or get it working properly before merging it.

msfschaffner reviewed (approved):
This looks good to me, and I think we should replicate this behavior in the other lint parser scripts as well (verilator, ascentlint).

Regarding the CI failure, this does not only affect OTP, but all DV environments due some unsupported constructs in the generated CSR FPV assertion file. See also my comment on the linked issue.

msfschaffner commented on hw/lint/tools/veriblelint/parse-lint-report.py:
```diff
@@ -62,9 +62,10 @@ def get_results(resdir):
+        ("errors", r"^.*Error: .*"),
         ("errors", r"^E .*"),
         ("errors", r"^F .*"),
+        ("errors", r".*: syntax error, rejected.*"),
```
this explicitly accounts for Verible syntax errors

msfschaffner commented on hw/formal/tools/csr_assert_gen/csr_assert_gen.core:
```diff
@@ -5,7 +5,18 @@ CAPI=2:
+      # support certain SVA language constrycts, causing it to throw syntax
+      # errors. Remove this once the issue has been addressed upstream.
+      - "! tool_veriblelint ? (lowrisc:fpv:csr_assert_gen_script)"
+      - "tool_veriblelint ? (lowrisc:fpv:csr_assert_gen_dummy)"
```
this might be a bit controversial, so let me know what you think. 
the other option is to modify each DV corefile which I kind of tried to avoid since then the changes are spread all over.

rswarbrick replied:
I think this looks like a good approach, especially as it should be temporary.

msfschaffner commented on hw/lint/tools/veriblelint/parse-lint-report.py:
```diff
@@ -70,10 +70,8 @@ def get_results(resdir):
-         # remove once this has been fixed in Edalize or in the corefile.
-         r"^(?!WARNING: Unknown item symbiyosis in section Target)WARNING: .*"
+         r"^(?!WARNING: Unknown item formal in section Target)"
+         r"WARNING: .*"
```
this basically repeats what we've done on this (reverted commit): https://github.com/lowRISC/opentitan/commit/9ce4bd9eba8ee99a301ce52f136ca70aa06ccf93

msfschaffner commented on hw/dv/tools/ralgen/ralgen.py:
```diff
@@ -60,11 +60,11 @@ def main():
         ral_spec = get_full_path(root_dir, top_hjson)
         cmd = os.path.join(util_path, "topgen.py")
-        args = [cmd, "-r", "-o", ".", "-t", ral_spec]
+        args = [cmd, "--nowarn", "-r", "-o", ".", "-t", ral_spec]
```
@sriyerg @cindychip 

this may not be the cleanest solution, but it does the trick.
let me know if this works for you or if you have a better idea.

rswarbrick replied:
I really dislike this. Let's instead try to get a warning-free build. I'm going to drop this patch from this PR (since it's got my name on it, I reserve the right to behave like a despot :-D), but will push it up as a draft PR so that it doesn't get lost. I'd prefer not to do this, though.

msfschaffner replied:
I didn't claim it was a pretty solution at all ;). 
we can also try to clean up the warning messages in topgen as much as possible and find a way to waive the remaining ones.

msfschaffner commented on hw/lint/tools/ascentlint/parse-lint-report.py:
```diff
@@ -32,38 +32,75 @@ def get_results(resdir):
     """
     results = {
         "tool": "ascentlint",
+        "fusesoc-error": [],
```
this last commit basically replicates what @rswarbrick has done in the first commit for the the Verible style lint parser

msfschaffner reviewed (commented):
Sorry for hijacking your PR, @rswarbrick, but I though this is probably the easiest way to get this unstuck. These are just suggestions - feel free to modify my commits.

rswarbrick commented:
Thanks for the improvements, @msfschaffner. I've just rebased and squashed together some commits. I've also split out the `--nowarn` commit into a separate draft PR #5042. I don't really think it's the right thing to do, but didn't want to unilaterally drop it without trace!

I'm happy with your changes that are left in this PR (so that's a review for half of it...). Could you check that you're happy with what I've done, and that I haven't done something terrible and left your name on it?

rswarbrick commented:
Ohh... Looking at the CI results, now I understand about the `--nowarn` flag! I'll push up some proposed changes to silence some of the warnings today (this PR isn't exactly urgent). Let's see where that gets us, and whether there's some stuff that we want to keep flying past on the console :-p

msfschaffner commented:
Thanks @rswarbrick, the rebased version LGTM.
Yeah this is not urgent, let me know if you find a better solution to handle the topgen warnings.

sriyerg commented on hw/lint/tools/ascentlint/parse-lint-report.py:
```diff
@@ -32,38 +32,75 @@ def get_results(resdir):
+    ])
+
+    # Patterns for ascentlint.log
+    err_warn_patterns.append([
```
@msfschaffner, how about we set these in the HJson instead?

rswarbrick replied:
I'm not sure whether that makes much difference: `ascentlint/parse-lint-report.py` already has to know stuff about AscentLint's output format, so I don't think it really matters to have some magic patterns in the code.

msfschaffner replied:
yeah that's right, the report parser scripts for these flows are pretty specific for the actual tool.
you could however argue that the lint parser script all follow the same structure (i.e., they look for the same message categories). hence, pulling these message patterns out into the hjson could allow us to unify all lint parser scripts into a generic one that can be reused among the lint flows.
I agree that this would be nice, but I would regard this as a cleanup/P3-ish step for the time being.

msfschaffner commented:
Thanks for getting this merged, @rswarbrick!
